### PR TITLE
Use triggers to restart the Agent whenever the configuration is modified

### PR DESCRIPTION
**Pull request by vboulineau**
What does this PR do?
---------------------

Use triggers to restart the Agent whenever the configuration is modified

Which scenarios this will impact?
-------------------

Docker

Motivation
----------

Additional Notes
----------------

### CODE REVIEW

File: aws/eks/eks/run.go [hunk 1]
```diff
@@ -76,10 +76,10 @@ func Run(ctx *pulumi.Context) error {
 		EndpointPublicAccess:         pulumi.BoolPtr(false),
 		Fargate:                      fargateProfile,
 		ClusterSecurityGroup:         clusterSG,
-		NodeAssociatePublicIpAddress: pulumi.BoolPtr(false),
+		NodeAssociatePublicIpAddress: pulumi.BoolRef(false),
```
Comment (pducolin) [inline aws/eks/eks/run.go:79]: **❓ question**
What is the difference between `BoolPtr` and `BoolRef` ? (Yes I could google the doc)

File: aws/ec2/ec2/vm.go [hunk 2]
```diff
@@ -21,8 +21,11 @@ type VM struct {
 	DockerManager     *command.DockerManager
 }
 
-func NewVM(ctx *pulumi.Context) (vm *VM, err error) {
-	vm.context = ctx
+func NewVM(ctx *pulumi.Context) (*VM, error) {
+	vm := &VM{
```
Comment (pducolin) [inline aws/ec2/ec2/vm.go:25]: Ouch, thank you !

File: command/docker.go [hunk 3]
```diff
@@ -61,14 +62,14 @@ func (d *DockerManager) ComposeFileUp(composeFilePath string, opts ...pulumi.Res
 
 	return d.runner.Command(
 		d.namer.ResourceName("run", composeFilePath),
-		pulumi.Sprintf("docker-compose -f %s up --detach --wait --timeout 300", remoteComposePath),
-		nil,
-		pulumi.Sprintf("docker-compose -f %s down -t 300", remoteComposePath),
-		nil, true,
+		&CommandArgs{
```
Comment (pducolin) [inline command/docker.go:65]: **👏 praise**
Love this !

File: command/docker.go [hunk 4]
```diff
@@ -61,14 +62,14 @@ func (d *DockerManager) ComposeFileUp(composeFilePath string, opts ...pulumi.Res
...
+			Create: pulumi.Sprintf("docker-compose -f %s up --detach --wait --timeout 300", remoteComposePath),
```
Comment (pducolin) [inline command/docker.go:66]: **🥜 nitpick**
Did you mean 
```suggestion
			Create: pulumi.Sprintf("docker-compose -f %s up --detach --wait --timeout %d", remoteComposePath, defaultTimeout),
```
?

File: command/docker.go [hunk 5]
```diff
@@ -61,14 +62,14 @@ func (d *DockerManager) ComposeFileUp(composeFilePath string, opts ...pulumi.Res
...
+			Delete: pulumi.Sprintf("docker-compose -f %s down -t 300", remoteComposePath),
+		},
 		pulumi.DependsOn([]pulumi.Resource{installCommand, copyCmd}))
 }
 
-func (d *DockerManager) ComposeStrUp(name string, composeManifests []DockerComposeInlineManifest, opts ...pulumi.ResourceOption) (*remote.Command, error) {
+func (d *DockerManager) ComposeStrUp(name string, composeManifests []DockerComposeInlineManifest, envVars pulumi.StringMap, opts ...pulumi.ResourceOption) (*remote.Command, error) {
```
Comment (pducolin) [inline command/docker.go:72]: **👏 praise**
😍

**Review by pducolin — APPROVED** [review]
Awesome !

Looking at command.Docker we could refactor the file and string compose to share some code, but that's for later

**Comment by pducolin** [comment]
**💬 suggestion**
I see code not related to the PR goal, suggestion for next PR to split in more commits to help review single changes

**Comment by pducolin** [comment]
**❓ question**

I don't see which change allows the agent restart on config change 🙈

File: aws/eks/eks/run.go (hunk 1 above)
Comment (vboulineau) [reply aws/eks/eks/run.go:79]: They changed the type in EKS provider 1.0.0, it was `pulumi.Bool` (`BoolPtr`), now it's just `*bool` (`BoolRef`)

**Comment by vboulineau** [comment]
> **❓ question**
> 
> I don't see which change allows the agent restart on config change 🙈

Look at the `Triggers` being field being filled with all content input.

File: command/docker.go [hunk 6]
```diff
@@ -80,27 +81,37 @@ func (d *DockerManager) ComposeStrUp(name string, composeManifests []DockerCompo
...
+				Create: utils.WriteStringCommand(manifest.Content, remoteComposePath),
+			},
+			pulumi.DependsOn([]pulumi.Resource{tempCmd}))
 		if err != nil {
 			return nil, err
 		}
-		writeCommands = append(writeCommands, writeCommand)
+
+		runCommandDeps = append(runCommandDeps, writeCommand)
+		runCommandTriggers = append(runCommandTriggers, manifest.Content)
```
[the remark this answers is not in the document]
Comment (vboulineau) [reply command/docker.go:101]: Correct

**Pull request merged**